File: 39.Combination_Sum.cpp
class Solution {
    public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> curr;
        combinationSumHelper(res, candidates, curr, target, 0);
        return res;
    }
    
    void combinationSumHelper(vector<vector<int>>& res, vector<int>& nums, vector<int>& curr, int target, int index){
        if (target == 0){
            res.push_back(curr);
            return;
        }
        else {
            for (int i=index; i<nums.size(); i++){
                if (target >= nums[i]){
                    curr.push_back(nums[i]);
                    combinationSumHelper(res, nums, curr, target-nums[i], i);
                    curr.pop_back();
                }
            }
        }
    }
};

File: 39.CombinationSum.py
class Solution:
    def __init__(self):
        self.res = []
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.backtracking([], target, 0, candidates)
        return self.res
        
    def backtracking(self, path, target, index, candidates):
        if target == 0:
            self.res.append(path)
        else:
            for i in range(index, len(candidates)):
                if candidates[i] <= target:
                    self.backtracking(path+[candidates[i]], target-candidates[i], i, candidates)
        return